### ISSUE TITLE
Reporting table sorting apparently done alphabetically

### ISSUE THREAD

**Issue by beeb** [autonomio/talos#193]
When calling `ta.Reporting(t).table('val_loss', sort_by='val_loss', ascending=True)` the sorting sometimes pust the first row at the very end of the table.

	  optimizer dense_nodes kernel_size filters            val_loss batch_size dropout
	4      adam           8           5       8   102.0532791929425          5     0.3
	3      adam           8           3       4  112.32208496669553         20     0.1
	1     nadam           8           5       8  120.11313240483122         40     0.3
	2     nadam          16           5       8   98.14560138054614         40     0.1

Similarly, `best_params()` would return the values for the first (wrong) row of the table instead of the last (correct) values.

Attached is the CSV file generated by Talos.

[021219092716_.zip](https://github.com/autonomio/talos/files/2854802/021219092716_.zip)

**Comment by beeb** [comment]
Could it be that sorting is done alphabetically?

**Comment by mikkokotila** [comment]
Using the file you have given, I've: 

```
import talos as ta
r = ta.Reporting('/Users/mikko/Downloads/021219092716_ 3.csv')
r.table('val_loss', sort_by='val_loss')
```
Unfortunately I can't replicate the issue you are reporting. 

Because I have not hear of this or anything similar in the past with Talos, my guess is something funky might be going on with the environment you are in. Can you create a [clean environment and install the latest (daily-dev) version of Talos](https://github.com/autonomio/talos/wiki/Installation-Options) and try to replicate the issue.

**Comment by beeb** [comment]
You are right that the issue does not appear when initializing the `Reporting` object with a csv file. In my case, I initialized it with the `Scan` object and that was the behavior. One particularity is the fact that my dictionary for the hyperparameters is an `OrderedDict`, could that have an influence?

**Comment by beeb** [comment]
So if I check `t.data.dtypes`, `t` being the `Scan` object, I get the dtype 'object' for all columns. That probably results in the sort function sorting as `str` instead of the correct numerical dtype.

**Comment by beeb** [comment]
Found this bit https://github.com/autonomio/talos/blob/25133d69568e06a6faa6e8071ce32c519e719d4e/talos/scan/scan_finish.py#L30

Are we sure that the resulting DataFrame has the correct dtypes? Seems like a string gets split here.

**Comment by beeb** [comment]
So after some testing, the sorting is definitely alphabetical if the dtype is `object`. I reproduced the steps that talos performs on the `self.result` object and that's what I get:

	>>> l = ['10,20,30,test', '4,5,6,test2']
	>>> df = pd.DataFrame(l)
	>>> df.dtypes
	0    object
	dtype: object
	>>> res = df[df.columns[0]].str.split(',', expand=True)
	>>> res
	    0   1   2      3
	0  10  20  30   test
	1   4   5   6  test2
	>>> res.sort_values(0, ascending=True)
	    0   1   2      3
	0  10  20  30   test
	1   4   5   6  test2

As you can see the sort puts 10 before 4, which means it's considered as a string.

**Comment by mikkokotila** [comment]
Good catch :) Because this has to do with the mainline of the program, the fix will take some time. I will allocate time for this later today, so might have an update later. It seems like a quick fix, but need to think about everything that might be effected.

**Comment by beeb** [comment]
Didn't have much success with `infer_objects` for now but [`pd.to_numeric()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_numeric.html#pandas.to_numeric) on each column as a `Series` seems to work.

**Comment by beeb** [comment]
Just sharing something that I encountered while implementing the fix on my end: if I `Scan().data.apply(pd.to_numeric, errors='ignore')` the result is correct but then I had to cast the values back to standard python types (with the `item()` method of numpy dtypes) because Keras complains about `np.generic` types for certain parameters (like `kernel_size` for convolutional layers).

**Comment by mikkokotila** [comment]
@beeb thanks :) Yes, this is a little bit tricky as I remember. 

I've made a commit https://github.com/autonomio/talos/commit/74163575e9402b39caf3c3c361c41161d76377a5 to daily-dev with a candidate for handling this. I did not "plug it in" yet, will have to do some testing. For the time being: 

```
from talos.utils.string_cols_to_numeric import string_cols_to_numeric
scan_object.data = string_cols_to_numeric(scan_object.data)
```
This appears to work. This can be executed somewhere after: https://github.com/autonomio/talos/blob/25133d69568e06a6faa6e8071ce32c519e719d4e/talos/scan/scan_finish.py#L30

That will then resolve all the other cases (e.g. Reporting).
